Node Upgrade: Change the upgrade script to keep os distro during upgrade.

Random-Liu opened the pull request:
This PR fixes https://github.com/kubernetes/kubernetes/issues/32812#issuecomment-247473584.

This PR will fix flakiness in upgrade tests for v1.4.0

Basically, our script didn't detect the image currently being used, it just use the [default image hard coded in the script](https://github.com/kubernetes/kubernetes/blob/d9ea0887653b62597a332d02453db53a4df2d494/cluster/gce/config-default.sh#L39) during upgrade.

This PR changes the script to detect the os distro currently being used and keep it unchanged during the upgrade.

@vishh @dchen1107
/cc @kubernetes/sig-cluster-lifecycle

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32840)

<!-- Reviewable:end -->

Random-Liu committed:
Change the upgrade script to keep os distro during upgrade.

vishh commented:
This change LGTM. We need to clean up the scripts to avoid auto sourcing deps later.

maisem commented on cluster/gce/upgrade.sh:
```diff
@@ -76,6 +76,9 @@ function upgrade-master() {
   detect-master
   parse-master-env
 
+  MASTER_OS_DISTRIBUTION=$(get-node-os "${MASTER_NAME}")
```
This should only keep the current os if no os is explicitly asked for.

maisem replied:
You could add it as a flag maybe or check if the env variable is already defined.

Random-Liu replied:
The env is always defined as default os distro, we just override it here. But yeah, we can add a flag.

maisem replied:
What I mean is something like
```shell
local -r current_master_os=$(get-node-os "${MASTER_NAME}")
MASTER_OS_DISTRIBUTION="${MASTER_OS_DISTRIBUTION:-${current_master_os}}"
```

maisem commented on cluster/gce/upgrade.sh:
```diff
@@ -177,6 +194,8 @@ function prepare-node-upgrade() {
   SANITIZED_VERSION=$(echo ${KUBE_VERSION} | sed 's/[\.\+]/-/g')
 
   detect-node-names # sets INSTANCE_GROUPS
+  NODE_OS_DISTRIBUTION=$(get-node-os "${NODE_NAMES[0]}")
```
same here.

maisem commented on cluster/gce/upgrade.sh:
```diff
@@ -33,11 +33,12 @@ source "${KUBE_ROOT}/cluster/kube-util.sh"
   echo "  -M:  Upgrade master only"
   echo "  -N:  Upgrade nodes only"
   echo "  -P:  Node upgrade prerequisites only (create a new instance template)"
+  echo "  -o:  Use os distro sepcified in NODE_OS_DISTRIBUTION for new nodes"
```
s/NODE_OS_DISTRIBUTION/KUBE_NODE_OS_DISTRIBUTION

maisem commented:
LGTM other than the minor nit.

Random-Liu commented:
I've tested the test for upgrade for v1.3.7 -> v1.4.0-beta.5

Before the upgrade
<details>
```shell
$ kubectl describe nodes kubernetes-master
Name:			kubernetes-master
Labels:			beta.kubernetes.io/arch=amd64
			beta.kubernetes.io/instance-type=n1-standard-1
			beta.kubernetes.io/os=linux
			failure-domain.beta.kubernetes.io/region=us-central1
			failure-domain.beta.kubernetes.io/zone=us-central1-b
			kubernetes.io/hostname=kubernetes-master
Taints:			<none>
CreationTimestamp:	Thu, 15 Sep 2016 20:58:16 -0700
Phase:			
Conditions:
  Type			Status	LastHeartbeatTime			LastTransitionTime			Reason				Message
  ----			------	-----------------			------------------			------				-------
  NetworkUnavailable 	False 	Mon, 01 Jan 0001 00:00:00 +0000 	Thu, 15 Sep 2016 20:58:36 -0700 	RouteCreated 			RouteController created a route
  OutOfDisk 		False 	Thu, 15 Sep 2016 20:59:37 -0700 	Thu, 15 Sep 2016 20:58:16 -0700 	KubeletHasSufficientDisk 	kubelet has sufficient disk space available
  MemoryPressure 	False 	Thu, 15 Sep 2016 20:59:37 -0700 	Thu, 15 Sep 2016 20:58:16 -0700 	KubeletHasSufficientMemory 	kubelet has sufficient memory available
  Ready 		True 	Thu, 15 Sep 2016 20:59:37 -0700 	Thu, 15 Sep 2016 20:58:16 -0700 	KubeletReady 			kubelet is posting ready status
Addresses:		10.240.0.5,104.197.66.111
Capacity:
 alpha.kubernetes.io/nvidia-gpu:	0
 cpu:					1
 memory:				3787412Ki
 pods:					110
Allocatable:
 alpha.kubernetes.io/nvidia-gpu:	0
 cpu:					1
 memory:				3787412Ki
 pods:					110
System Info:
 Machine ID:			7a11aa832f88892f25df79de57db6d7a
 System UUID:			2F1C42F3-6261-2B63-23AC-6741DBF27C16
 Boot ID:			31cc6122-6711-47aa-97e2-9ef57c2f8bad
 Kernel Version:		4.4.4+
 OS Image:			Google Container-VM Image
 Operating System:		linux
 Architecture:			amd64
 Container Runtime Version:	docker://1.11.2
 Kubelet Version:		v1.3.7
 Kube-Proxy Version:		v1.3.7
PodCIDR:			10.244.0.0/24
ExternalID:			8391252370720740231
Non-terminated Pods:		(10 in total)
  Namespace			Name							CPU Requests	CPU Limits	Memory Requests	Memory Limits
  ---------			----							------------	----------	---------------	-------------
  kube-system			etcd-empty-dir-cleanup-kubernetes-master		0 (0%)		0 (0%)		0 (0%)		0 (0%)
  kube-system			etcd-server-events-kubernetes-master			100m (10%)	0 (0%)		0 (0%)		0 (0%)
  kube-system			etcd-server-kubernetes-master				200m (20%)	0 (0%)		0 (0%)		0 (0%)
  kube-system			fluentd-cloud-logging-kubernetes-master			80m (8%)	0 (0%)		200Mi (5%)	200Mi (5%)
  kube-system			kube-addon-manager-kubernetes-master			5m (0%)		0 (0%)		50Mi (1%)	0 (0%)
  kube-system			kube-apiserver-kubernetes-master			250m (25%)	0 (0%)		0 (0%)		0 (0%)
  kube-system			kube-controller-manager-kubernetes-master		200m (20%)	0 (0%)		0 (0%)		0 (0%)
  kube-system			kube-scheduler-kubernetes-master			100m (10%)	0 (0%)		0 (0%)		0 (0%)
  kube-system			l7-lb-controller-v0.7.0-kubernetes-master		10m (1%)	100m (10%)	50Mi (1%)	100Mi (2%)
  kube-system			node-problem-detector-v0.1-de65n			20m (2%)	200m (20%)	20Mi (0%)	100Mi (2%)
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted. More info: http://releases.k8s.io/HEAD/docs/user-guide/compute-resources.md)
  CPU Requests	CPU Limits	Memory Requests	Memory Limits
  ------------	----------	---------------	-------------
  965m (96%)	300m (30%)	320Mi (8%)	400Mi (10%)
Events:
  FirstSeen	LastSeen	Count	From				SubobjectPath	Type		Reason			Message
  ---------	--------	-----	----				-------------	--------	------			-------
  2m		2m		1	{kubelet kubernetes-master}			Normal		Starting		Starting kubelet.
  2m		2m		1	{kubelet kubernetes-master}			Normal		NodeNotSchedulable	Node kubernetes-master status is now: NodeNotSchedulable
  2m		1m		47	{kubelet kubernetes-master}			Normal		NodeHasSufficientDisk	Node kubernetes-master status is now: NodeHasSufficientDisk
  2m		1m		47	{kubelet kubernetes-master}			Normal		NodeHasSufficientMemory	Node kubernetes-master status is now: NodeHasSufficientMemory
  1m		1m		1	{controllermanager }				Normal		RegisteredNode		Node kubernetes-master event: Registered Node kubernetes-master in NodeController

$ kubectl describe nodes kubernetes-minion-group-1ifk
Name:			kubernetes-minion-group-1ifk
Labels:			beta.kubernetes.io/arch=amd64
			beta.kubernetes.io/instance-type=n1-standard-2
			beta.kubernetes.io/os=linux
			failure-domain.beta.kubernetes.io/region=us-central1
			failure-domain.beta.kubernetes.io/zone=us-central1-b
			kubernetes.io/hostname=kubernetes-minion-group-1ifk
Taints:			<none>
CreationTimestamp:	Thu, 15 Sep 2016 20:58:26 -0700
Phase:			
Conditions:
  Type			Status	LastHeartbeatTime			LastTransitionTime			Reason				Message
  ----			------	-----------------			------------------			------				-------
  NetworkUnavailable 	False 	Mon, 01 Jan 0001 00:00:00 +0000 	Thu, 15 Sep 2016 20:59:10 -0700 	RouteCreated 			RouteController created a route
  OutOfDisk 		False 	Thu, 15 Sep 2016 20:59:43 -0700 	Thu, 15 Sep 2016 20:58:26 -0700 	KubeletHasSufficientDisk 	kubelet has sufficient disk space available
  MemoryPressure 	False 	Thu, 15 Sep 2016 20:59:43 -0700 	Thu, 15 Sep 2016 20:58:26 -0700 	KubeletHasSufficientMemory 	kubelet has sufficient memory available
  Ready 		True 	Thu, 15 Sep 2016 20:59:43 -0700 	Thu, 15 Sep 2016 20:59:03 -0700 	KubeletReady 			kubelet is posting ready status. WARNING: CPU hardcapping unsupported
Addresses:		10.240.0.7,104.197.239.59
Capacity:
 alpha.kubernetes.io/nvidia-gpu:	0
 cpu:					2
 memory:				7679820Ki
 pods:					110
Allocatable:
 alpha.kubernetes.io/nvidia-gpu:	0
 cpu:					2
 memory:				7679820Ki
 pods:					110
System Info:
 Machine ID:			
 System UUID:			9CDF8CDD-EF72-526D-F00E-2F4C154532DE
 Boot ID:			7ea96bae-6f31-4001-b1c6-16ed3f3e80de
 Kernel Version:		3.16.0-4-amd64
 OS Image:			Debian GNU/Linux 7 (wheezy)
 Operating System:		linux
 Architecture:			amd64
 Container Runtime Version:	docker://1.11.2
 Kubelet Version:		v1.3.7
 Kube-Proxy Version:		v1.3.7
PodCIDR:			10.244.2.0/24
ExternalID:			2655602165736199039
Non-terminated Pods:		(3 in total)
  Namespace			Name								CPU Requests	CPU Limits	Memory Requests	Memory Limits
  ---------			----								------------	----------	---------------	-------------
  kube-system			fluentd-cloud-logging-kubernetes-minion-group-1ifk		100m (5%)	0 (0%)		200Mi (2%)	200Mi (2%)
  kube-system			kube-proxy-kubernetes-minion-group-1ifk				100m (5%)	0 (0%)		0 (0%)		0 (0%)
  kube-system			node-problem-detector-v0.1-zjdx9				20m (1%)	200m (10%)	20Mi (0%)	100Mi (1%)
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted. More info: http://releases.k8s.io/HEAD/docs/user-guide/compute-resources.md)
  CPU Requests	CPU Limits	Memory Requests	Memory Limits
  ------------	----------	---------------	-------------
  220m (11%)	200m (10%)	220Mi (2%)	300Mi (4%)
Events:
  FirstSeen	LastSeen	Count	From						SubobjectPath	Type		Reason			Message
  ---------	--------	-----	----						-------------	--------	------			-------
  1m		1m		1	{kubelet kubernetes-minion-group-1ifk}				Normal		Starting		Starting kubelet.
  1m		1m		1	{kubelet kubernetes-minion-group-1ifk}				Normal		NodeHasSufficientDisk	Node kubernetes-minion-group-1ifk status is now: NodeHasSufficientDisk
  1m		1m		1	{kubelet kubernetes-minion-group-1ifk}				Normal		NodeHasSufficientMemory	Node kubernetes-minion-group-1ifk status is now: NodeHasSufficientMemory
  1m		1m		1	{controllermanager }						Normal		RegisteredNode		Node kubernetes-minion-group-1ifk event: Registered Node kubernetes-minion-group-1ifk in NodeController
  1m		1m		1	{kubelet kubernetes-minion-group-1ifk}				Normal		Starting		Starting kubelet.
  1m		1m		1	{kubelet kubernetes-minion-group-1ifk}				Normal		NodeHasSufficientDisk	Node kubernetes-minion-group-1ifk status is now: NodeHasSufficientDisk
  1m		1m		1	{kubelet kubernetes-minion-group-1ifk}				Normal		NodeHasSufficientMemory	Node kubernetes-minion-group-1ifk status is now: NodeHasSufficientMemory
  45s		45s		1	{kubelet kubernetes-minion-group-1ifk}				Normal		NodeReady		Node kubernetes-minion-group-1ifk status is now: NodeReady
  39s		39s		1	{kube-proxy kubernetes-minion-group-1ifk}			Normal		Starting		Starting kube-proxy.
```
</details>

Upgrade Master:
<detail>
```shell
$ cluster/gce/upgrade.sh -M v1.4.0-beta.5
Your active configuration is: [default]

Project: google.com:noogler-kubernetes
Zone: us-central1-b
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 45515  100 45515    0     0   367k      0 --:--:-- --:--:-- --:--:--  370k
== Upgrading master to 'https://storage.googleapis.com/kubernetes-release/release/v1.4.0-beta.5/kubernetes-server-linux-amd64.tar.gz'. Do not interrupt, deleting master instance. ==
Trying to find master named 'kubernetes-master'
Looking for address 'kubernetes-master-ip'
Using master: kubernetes-master (external IP: 104.197.66.111)
Deleted [https://www.googleapis.com/compute/v1/projects/google.com:noogler-kubernetes/zones/us-central1-b/instances/kubernetes-master].
WARNING: You have selected a disk size of under [200GB]. This may result in poor I/O performance. For more information, see: https://developers.google.com/compute/docs/disks#pdperformance.
Created [https://www.googleapis.com/compute/v1/projects/google.com:noogler-kubernetes/zones/us-central1-b/instances/kubernetes-master].
NAME               ZONE           MACHINE_TYPE   PREEMPTIBLE  INTERNAL_IP  EXTERNAL_IP     STATUS
kubernetes-master  us-central1-b  n1-standard-1               10.240.0.5   104.197.66.111  RUNNING
== Waiting for new master to respond to API requests ==
................................................== Done ==
== Validating cluster post-upgrade ==
Found 4 node(s).
NAME                           STATUS                     AGE
kubernetes-master              Ready,SchedulingDisabled   5m
kubernetes-minion-group-1ifk   Ready                      5m
kubernetes-minion-group-iwh2   Ready                      5m
kubernetes-minion-group-okad   Ready                      5m
Validate output:
NAME                 STATUS    MESSAGE              ERROR
controller-manager   Healthy   ok                   
scheduler            Healthy   ok                   
etcd-0               Healthy   {"health": "true"}   
etcd-1               Healthy   {"health": "true"}   
Cluster validation succeeded
```
</detail>

Upgrade Node:
<detail>
```
$ cluster/gce/upgrade.sh -N v1.4.0-beta.5
Your active configuration is: [default]

Project: google.com:noogler-kubernetes
Zone: us-central1-b
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 45515  100 45515    0     0   395k      0 --:--:-- --:--:-- --:--:--  396k
== Preparing node upgrade (to v1.4.0-beta.5). ==
INSTANCE_GROUPS=kubernetes-minion-group
NODE_NAMES=kubernetes-minion-group-1ifk kubernetes-minion-group-iwh2 kubernetes-minion-group-okad
Attempt 1 to create kubernetes-minion-template-v1-4-0-beta-5
WARNING: You have selected a disk size of under [200GB]. This may result in poor I/O performance. For more information, see: https://developers.google.com/compute/docs/disks#pdperformance.
Created [https://www.googleapis.com/compute/v1/projects/google.com:noogler-kubernetes/global/instanceTemplates/kubernetes-minion-template-v1-4-0-beta-5].
NAME                                      MACHINE_TYPE   PREEMPTIBLE  CREATION_TIMESTAMP
kubernetes-minion-template-v1-4-0-beta-5  n1-standard-2               2016-09-15T21:05:11.717-07:00
Instance template name: kubernetes-minion-template-v1-4-0-beta-5
== Finished preparing node upgrade (to v1.4.0-beta.5). ==
== Upgrading nodes to v1.4.0-beta.5. ==
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is still in ROLLING_FORWARD state.
Rolling update 459e2b05-3f58-440e-8a62-8d8018908d8b is ROLLED_OUT state and finished.
Deleted [https://www.googleapis.com/compute/v1/projects/google.com:noogler-kubernetes/global/instanceTemplates/kubernetes-minion-template].
== Finished upgrading nodes to v1.4.0-beta.5. ==
== Validating cluster post-upgrade ==
Waiting for 4 ready nodes. 3 ready nodes, 4 registered. Retrying.
Waiting for 4 ready nodes. 3 ready nodes, 4 registered. Retrying.
Waiting for 4 ready nodes. 3 ready nodes, 4 registered. Retrying.
Waiting for 4 ready nodes. 3 ready nodes, 4 registered. Retrying.
Waiting for 4 ready nodes. 3 ready nodes, 4 registered. Retrying.
Waiting for 4 ready nodes. 3 ready nodes, 4 registered. Retrying.
Found 4 node(s).
NAME                           STATUS                     AGE
kubernetes-master              Ready,SchedulingDisabled   8m
kubernetes-minion-group-1ifk   Ready                      3m
kubernetes-minion-group-iwh2   Ready                      2m
kubernetes-minion-group-okad   Ready                      50s
Validate output:
NAME                 STATUS    MESSAGE              ERROR
controller-manager   Healthy   ok                   
scheduler            Healthy   ok                   
etcd-0               Healthy   {"health": "true"}   
etcd-1               Healthy   {"health": "true"}   
Cluster validation succeeded
```
</detail>

After Upgrade:
<detail>
```
$ kubectl describe nodes kubernetes-master
Name:			kubernetes-master
Labels:			beta.kubernetes.io/arch=amd64
			beta.kubernetes.io/instance-type=n1-standard-1
			beta.kubernetes.io/os=linux
			failure-domain.beta.kubernetes.io/region=us-central1
			failure-domain.beta.kubernetes.io/zone=us-central1-b
			kubernetes.io/hostname=kubernetes-master
Taints:			<none>
CreationTimestamp:	Thu, 15 Sep 2016 21:04:20 -0700
Phase:			
Conditions:
  Type			Status	LastHeartbeatTime			LastTransitionTime			Reason				Message
  ----			------	-----------------			------------------			------				-------
  NetworkUnavailable 	False 	Thu, 15 Sep 2016 21:20:33 -0700 	Thu, 15 Sep 2016 21:20:33 -0700 	RouteCreated 			RouteController created a route
  OutOfDisk 		False 	Thu, 15 Sep 2016 21:20:33 -0700 	Thu, 15 Sep 2016 21:04:20 -0700 	KubeletHasSufficientDisk 	kubelet has sufficient disk space available
  MemoryPressure 	False 	Thu, 15 Sep 2016 21:20:33 -0700 	Thu, 15 Sep 2016 21:04:20 -0700 	KubeletHasSufficientMemory 	kubelet has sufficient memory available
  DiskPressure 		False 	Thu, 15 Sep 2016 21:20:33 -0700 	Thu, 15 Sep 2016 21:04:20 -0700 	KubeletHasNoDiskPressure 	kubelet has no disk pressure
  Ready 		True 	Thu, 15 Sep 2016 21:20:33 -0700 	Thu, 15 Sep 2016 21:04:20 -0700 	KubeletReady 			kubelet is posting ready status. AppArmor enabled
Addresses:		10.240.0.5,104.197.66.111
Capacity:
 alpha.kubernetes.io/nvidia-gpu:	0
 cpu:					1
 memory:				3788348Ki
 pods:					110
Allocatable:
 alpha.kubernetes.io/nvidia-gpu:	0
 cpu:					1
 memory:				3788348Ki
 pods:					110
System Info:
 Machine ID:			ba0e80dabd33e222bb19257f57db6eb3
 System UUID:			2F1C42F3-6261-2B63-23AC-6741DBF27C16
 Boot ID:			263d2da1-3bd8-4def-989c-a878d94fd395
 Kernel Version:		4.4.14+
 OS Image:			Google Container-VM Image
 Operating System:		linux
 Architecture:			amd64
 Container Runtime Version:	docker://1.11.2
 Kubelet Version:		v1.4.0-beta.5
 Kube-Proxy Version:		v1.4.0-beta.5
PodCIDR:			10.244.0.0/24
ExternalID:			3887833226465955917
Non-terminated Pods:		(11 in total)
  Namespace			Name							CPU Requests	CPU Limits	Memory Requests	Memory Limits
  ---------			----							------------	----------	---------------	-------------
  kube-system			etcd-empty-dir-cleanup-kubernetes-master		0 (0%)		0 (0%)		0 (0%)		0 (0%)
  kube-system			etcd-server-events-kubernetes-master			100m (10%)	0 (0%)		0 (0%)		0 (0%)
  kube-system			etcd-server-kubernetes-master				200m (20%)	0 (0%)		0 (0%)		0 (0%)
  kube-system			fluentd-cloud-logging-kubernetes-master			80m (8%)	0 (0%)		200Mi (5%)	200Mi (5%)
  kube-system			kube-addon-manager-kubernetes-master			5m (0%)		0 (0%)		50Mi (1%)	0 (0%)
  kube-system			kube-apiserver-kubernetes-master			250m (25%)	0 (0%)		0 (0%)		0 (0%)
  kube-system			kube-controller-manager-kubernetes-master		200m (20%)	0 (0%)		0 (0%)		0 (0%)
  kube-system			kube-scheduler-kubernetes-master			100m (10%)	0 (0%)		0 (0%)		0 (0%)
  kube-system			l7-lb-controller-v0.7.1-kubernetes-master		10m (1%)	0 (0%)		50Mi (1%)	0 (0%)
  kube-system			node-problem-detector-v0.1-de65n			20m (2%)	200m (20%)	20Mi (0%)	100Mi (2%)
  kube-system			rescheduler-v0.2.1-kubernetes-master			10m (1%)	0 (0%)		100Mi (2%)	0 (0%)
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted. More info: http://releases.k8s.io/HEAD/docs/user-guide/compute-resources.md)
  CPU Requests	CPU Limits	Memory Requests	Memory Limits
  ------------	----------	---------------	-------------
  975m (97%)	200m (20%)	420Mi (11%)	300Mi (8%)
Events:
  FirstSeen	LastSeen	Count	From				SubobjectPath	Type		Reason			Message
  ---------	--------	-----	----				-------------	--------	------			-------
  22m		22m		1	{kubelet kubernetes-master}			Normal		Starting		Starting kubelet.
  22m		22m		1	{kubelet kubernetes-master}			Normal		NodeNotSchedulable	Node kubernetes-master status is now: NodeNotSchedulable
  22m		22m		47	{kubelet kubernetes-master}			Normal		NodeHasSufficientDisk	Node kubernetes-master status is now: NodeHasSufficientDisk
  22m		22m		47	{kubelet kubernetes-master}			Normal		NodeHasSufficientMemory	Node kubernetes-master status is now: NodeHasSufficientMemory
  22m		22m		1	{controllermanager }				Normal		RegisteredNode		Node kubernetes-master event: Registered Node kubernetes-master in NodeController
  17m		17m		1	{kubelet kubernetes-master}			Normal		Starting		Starting kubelet.
  17m		17m		1	{kubelet kubernetes-master}			Normal		NodeNotSchedulable	Node kubernetes-master status is now: NodeNotSchedulable
  17m		16m		105	{kubelet kubernetes-master}			Normal		NodeHasSufficientDisk	Node kubernetes-master status is now: NodeHasSufficientDisk
  17m		16m		105	{kubelet kubernetes-master}			Normal		NodeHasSufficientMemory	Node kubernetes-master status is now: NodeHasSufficientMemory
  17m		16m		105	{kubelet kubernetes-master}			Normal		NodeHasNoDiskPressure	Node kubernetes-master status is now: NodeHasNoDiskPressure

$ kubectl describe nodes kubernetes-minion-group-1ifk
Name:			kubernetes-minion-group-1ifk
Labels:			beta.kubernetes.io/arch=amd64
			beta.kubernetes.io/instance-type=n1-standard-2
			beta.kubernetes.io/os=linux
			failure-domain.beta.kubernetes.io/region=us-central1
			failure-domain.beta.kubernetes.io/zone=us-central1-b
			kubernetes.io/hostname=kubernetes-minion-group-1ifk
Taints:			<none>
CreationTimestamp:	Thu, 15 Sep 2016 21:08:48 -0700
Phase:			
Conditions:
  Type			Status	LastHeartbeatTime			LastTransitionTime			Reason				Message
  ----			------	-----------------			------------------			------				-------
  NetworkUnavailable 	False 	Thu, 15 Sep 2016 21:20:53 -0700 	Thu, 15 Sep 2016 21:20:53 -0700 	RouteCreated 			RouteController created a route
  OutOfDisk 		False 	Thu, 15 Sep 2016 21:20:56 -0700 	Thu, 15 Sep 2016 21:08:47 -0700 	KubeletHasSufficientDisk 	kubelet has sufficient disk space available
  MemoryPressure 	False 	Thu, 15 Sep 2016 21:20:56 -0700 	Thu, 15 Sep 2016 21:08:47 -0700 	KubeletHasSufficientMemory 	kubelet has sufficient memory available
  DiskPressure 		False 	Thu, 15 Sep 2016 21:20:56 -0700 	Thu, 15 Sep 2016 21:08:47 -0700 	KubeletHasNoDiskPressure 	kubelet has no disk pressure
  Ready 		True 	Thu, 15 Sep 2016 21:20:56 -0700 	Thu, 15 Sep 2016 21:09:25 -0700 	KubeletReady 			kubelet is posting ready status. WARNING: CPU hardcapping unsupported
  KernelDeadlock 	False 	Thu, 15 Sep 2016 21:20:55 -0700 	Thu, 15 Sep 2016 21:09:32 -0700 	KernelHasNoDeadlock 		kernel has no deadlock
Addresses:		10.240.0.7,104.198.207.75
Capacity:
 alpha.kubernetes.io/nvidia-gpu:	0
 cpu:					2
 memory:				7679820Ki
 pods:					110
Allocatable:
 alpha.kubernetes.io/nvidia-gpu:	0
 cpu:					2
 memory:				7679820Ki
 pods:					110
System Info:
 Machine ID:			
 System UUID:			9CDF8CDD-EF72-526D-F00E-2F4C154532DE
 Boot ID:			63141e97-d41b-4e98-92d8-f3270c0d36c2
 Kernel Version:		3.16.0-4-amd64
 OS Image:			Debian GNU/Linux 7 (wheezy)
 Operating System:		linux
 Architecture:			amd64
 Container Runtime Version:	docker://1.11.2
 Kubelet Version:		v1.4.0-beta.5
 Kube-Proxy Version:		v1.4.0-beta.5
PodCIDR:			10.244.2.0/24
ExternalID:			3576005157528719651
Non-terminated Pods:		(3 in total)
  Namespace			Name								CPU Requests	CPU Limits	Memory Requests	Memory Limits
  ---------			----								------------	----------	---------------	-------------
  kube-system			fluentd-cloud-logging-kubernetes-minion-group-1ifk		100m (5%)	0 (0%)		200Mi (2%)	200Mi (2%)
  kube-system			kube-proxy-kubernetes-minion-group-1ifk				100m (5%)	0 (0%)		0 (0%)		0 (0%)
  kube-system			node-problem-detector-v0.1-51a7c				20m (1%)	200m (10%)	20Mi (0%)	100Mi (1%)
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted. More info: http://releases.k8s.io/HEAD/docs/user-guide/compute-resources.md)
  CPU Requests	CPU Limits	Memory Requests	Memory Limits
  ------------	----------	---------------	-------------
  220m (11%)	200m (10%)	220Mi (2%)	300Mi (4%)
Events:
  FirstSeen	LastSeen	Count	From						SubobjectPath	Type		Reason			Message
  ---------	--------	-----	----						-------------	--------	------			-------
  22m		22m		1	{kubelet kubernetes-minion-group-1ifk}				Normal		NodeHasSufficientDisk	Node kubernetes-minion-group-1ifk status is now: NodeHasSufficientDisk
  22m		22m		1	{kubelet kubernetes-minion-group-1ifk}				Normal		NodeHasSufficientMemory	Node kubernetes-minion-group-1ifk status is now: NodeHasSufficientMemory
  22m		22m		1	{kubelet kubernetes-minion-group-1ifk}				Normal		Starting		Starting kubelet.
  22m		22m		1	{controllermanager }						Normal		RegisteredNode		Node kubernetes-minion-group-1ifk event: Registered Node kubernetes-minion-group-1ifk in NodeController
  22m		22m		1	{kubelet kubernetes-minion-group-1ifk}				Normal		Starting		Starting kubelet.
  22m		22m		1	{kubelet kubernetes-minion-group-1ifk}				Normal		NodeHasSufficientMemory	Node kubernetes-minion-group-1ifk status is now: NodeHasSufficientMemory
  22m		22m		1	{kubelet kubernetes-minion-group-1ifk}				Normal		NodeHasSufficientDisk	Node kubernetes-minion-group-1ifk status is now: NodeHasSufficientDisk
  21m		21m		1	{kubelet kubernetes-minion-group-1ifk}				Normal		NodeReady		Node kubernetes-minion-group-1ifk status is now: NodeReady
  21m		21m		1	{kube-proxy kubernetes-minion-group-1ifk}			Normal		Starting		Starting kube-proxy.
  12m		12m		2	{kubelet kubernetes-minion-group-1ifk}				Normal		NodeHasNoDiskPressure	Node kubernetes-minion-group-1ifk status is now: NodeHasNoDiskPressure
  12m		12m		2	{kubelet kubernetes-minion-group-1ifk}				Normal		NodeHasSufficientDisk	Node kubernetes-minion-group-1ifk status is now: NodeHasSufficientDisk
  12m		12m		2	{kubelet kubernetes-minion-group-1ifk}				Normal		NodeHasSufficientMemory	Node kubernetes-minion-group-1ifk status is now: NodeHasSufficientMemory
  12m		12m		1	{kubelet kubernetes-minion-group-1ifk}				Normal		Starting		Starting kubelet.
  12m		12m		1	{kubelet kubernetes-minion-group-1ifk}				Normal		Starting		Starting kubelet.
  12m		12m		2	{kubelet kubernetes-minion-group-1ifk}				Normal		NodeHasSufficientDisk	Node kubernetes-minion-group-1ifk status is now: NodeHasSufficientDisk
  12m		12m		2	{kubelet kubernetes-minion-group-1ifk}				Normal		NodeHasSufficientMemory	Node kubernetes-minion-group-1ifk status is now: NodeHasSufficientMemory
  12m		12m		2	{kubelet kubernetes-minion-group-1ifk}				Normal		NodeHasNoDiskPressure	Node kubernetes-minion-group-1ifk status is now: NodeHasNoDiskPressure
  11m		11m		1	{kubelet kubernetes-minion-group-1ifk}				Normal		NodeReady		Node kubernetes-minion-group-1ifk status is now: NodeReady
  11m		11m		1	{kube-proxy kubernetes-minion-group-1ifk}			Normal		Starting		Starting kube-proxy.
```
</detail>

We can see that after the test, node is still container-VM.

I'll try downgrade later, but because it is using the same logic, it should also work.

This should be cherry-picked in v1.3 and v1.4.

Random-Liu commented:
In e2e, the error is:
```
Your current Cloud SDK version is: 126.0.0
Installing components from version: 126.0.0

+----------------------------------------------------------------------------+
|                    These components will be installed.                     |
+-----------------------------------------------------+------------+---------+
|                         Name                        |  Version   |   Size  |
+-----------------------------------------------------+------------+---------+
| BigQuery Command Line Tool                          |     2.0.24 | < 1 MiB |
| BigQuery Command Line Tool (Platform Specific)      |     2.0.24 | < 1 MiB |
| Cloud SDK Core Libraries (Platform Specific)        | 2016.03.28 | < 1 MiB |
| Cloud Storage Command Line Tool                     |       4.21 | 2.8 MiB |
| Cloud Storage Command Line Tool (Platform Specific) |       4.18 | < 1 MiB |
| Default set of gcloud commands                      |            |         |
+-----------------------------------------------------+------------+---------+

For the latest full release notes, please visit:
  https://cloud.google.com/sdk/release_notes

#============================================================#
#= Creating update staging area                             =#
#============================================================#
#= Installing: BigQuery Command Line Tool                   =#
#============================================================#
#= Installing: BigQuery Command Line Tool (Platform Spec... =#
#============================================================#
#= Installing: Cloud SDK Core Libraries (Platform Specific) =#
#============================================================#
ERROR: (gcloud.components.update) The component [core-nix] failed to download.

HTTP Error 503: Service Unavailable

This will install all the core command line tools necessary for working with
the Google Cloud Platform.

+++ exit code: 1
+++ error: 1
```

Random-Liu commented:
In integration test, it's the flake https://github.com/kubernetes/kubernetes/issues/32770.

Random-Liu commented:
@k8s-bot test this issue https://github.com/kubernetes/kubernetes/issues/32770.

davidopp commented:
This hit the gcloud api outage; I'll re-trigger test.

davidopp commented:
Well, some of the presubmits are still running so I guess I won't touch it for now.

vishh commented:
@k8s-bot kubemark e2e test this github issue #IGNORE

vishh commented:
Manually merging this PR to facilitate upgrade testing for release v1.4.0
